lychee-lib/src/client.rs

```diff
@@ -24,132 +33,194 @@ use crate::{
 pub const DEFAULT_RETRY_WAIT_TIME: u64 = 1;
-/// Total timeout per request until a request gets declared as failed
+/// Default timeout in seconds before a request is deemed as failed, 20.
 pub const DEFAULT_TIMEOUT: usize = 20;
+/// Default user agent, `lychee-<PKG_VERSION>`.
+pub const DEFAULT_USER_AGENT: &str = concat!("lychee/", env!("CARGO_PKG_VERSION"));
 
-/// Handles incoming requests and returns responses. Usually you would not
-/// initialize a `Client` yourself, but use the `ClientBuilder` because it
-/// provides sane defaults for all configuration options.
-#[derive(Debug, Clone)]
-pub struct Client {
-    /// Underlying reqwest client instance that handles the HTTP requests.
-    reqwest_client: reqwest::Client,
-    /// Github client.
-    github_client: Option<Github>,
-    /// Filtered domain handling.
-    filter: Filter,
-    /// Maximum number of retries
-    max_retries: u64,
-    /// Default request HTTP method to use.
-    method: reqwest::Method,
-    /// The set of accepted HTTP status codes for valid URIs.
-    accepted: Option<HashSet<StatusCode>>,
-    /// Require HTTPS URL when it's available.
-    require_https: bool,
-    /// Override behavior for certain known issues with URIs.
-    quirks: Quirks,
-}
-
-/// A link checker using an API token for Github links
-/// otherwise a normal HTTP client.
-#[allow(unreachable_pub)]
-#[derive(TypedBuilder, Debug)]
+/// Builder for [`Client`].
+///
+/// See crate-level documentation for usage example.
+#[derive(TypedBuilder, Debug, Clone)]
 #[builder(field_defaults(default, setter(into)))]
+#[builder(builder_method_doc = "
+Create a builder for building `ClientBuilder`.
+
+On the builder call, call methods with same name as its fields to set their values.
+
+Finally, call `.build()` to create the instance of `ClientBuilder`.
+")]
 pub struct ClientBuilder {
-    /// Set an optional Github token.
-    /// This allows for more requests before
-    /// getting rate-limited.
+    /// Optional Github token used for Github links.
+    ///
+    /// This allows much more request before getting rate-limited.
+    ///
+    /// *P.S.* As of Feb 2022, it's 60 per hour without Github token
+    /// v.s. 5000 per hour with token.
     github_token: Option<String>,
-    /// Check links matching this set of regular expressions
+    /// Links matching this set of regular expressions are **always** checked.
+    ///
+    /// This has higher precedence over [`ClientBuilder::excludes`], **but**
+    /// has lower precedence over any other `exclude_` fields or
+    /// [`ClientBuilder::schemes`] below.
     includes: Option<RegexSet>,
-    /// Exclude links matching this set of regular expressions
+    /// Links matching this set of regular expressions are ignored, **except**
+    /// when a link also matches against [`ClientBuilder::includes`].
     excludes: Option<RegexSet>,
-    /// Exclude all private network addresses
+    /// When it's `true`, exclude all private network addresses.
```

Comment:
I'd say "When true" instead of "When it's true", but it's more of a stylistic preference.

Reply:
Okay I will change that.